Fix binds logging for "WHERE ... IN ..." statements

ricardotk002 opened the pull request:
After the new node `Arel::Nodes::HomogeneousIn` was introduced in
72fd0ba, the way bindings are printed by the log subscriber was left
in a kind of broken state:

Rails 6.0.3 example output

```
User Load (0.7ms)  SELECT "users".* FROM "users" WHERE "users"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9)  [["id", 1], ["id", 2], ["id", 3], ["id", 4], ["id", 5], ["id", 6], ["id", 7], ["id", 8], ["id", 9]]
```

Rails 6.1.0 example output

```
User Load (0.5ms)  SELECT "users".* FROM "users" WHERE "users"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9)  [[nil, 1], [nil, 2], [nil, 3], [nil, 4], [nil, 5], [nil, 6], [nil, 7], [nil, 8], [nil, 9]]
```

The original implementation relied on the predicate builder to create
a list of `BindParam` nodes, each with a `QueryAttribute` used to wrap
the attribute name and value, whereas the new implementation just
passes an array of casted values as the binds.

This patch does a lighter version of that process by applying a proc
to the binds right before adding them to the collectors. Credits to @kamipo.

```ruby
ids = (1..10000).each.map do |n|
  Post.create!.id
end

Benchmark.ips do |x|
  x.report("where with ids") do
    Post.where(id: ids).to_a
  end

  x.report("where with sanitize") do
    Post.where(ActiveRecord::Base.sanitize_sql(["id IN (?)", ids])).to_a
  end

  x.compare!
end
```

Before patch (main)
```
Warming up --------------------------------------
      where with ids     1.000  i/100ms
 where with sanitize     1.000  i/100ms
Calculating -------------------------------------
      where with ids     12.097  (± 0.0%) i/s -     61.000  in   5.047874s
 where with sanitize     13.657  (± 0.0%) i/s -     69.000  in   5.057082s

Comparison:
 where with sanitize:       13.7 i/s
      where with ids:       12.1 i/s - 1.13x  (± 0.00) slower
```

After patch
```
Warming up --------------------------------------
      where with ids     1.000  i/100ms
 where with sanitize     1.000  i/100ms
Calculating -------------------------------------
      where with ids     12.346  (± 8.1%) i/s -     62.000  in   5.030338s
 where with sanitize     13.885  (± 7.2%) i/s -     70.000  in   5.051146s

Comparison:
 where with sanitize:       13.9 i/s
      where with ids:       12.3 i/s - same-ish: difference falls within error
```

<details>
  <summary>Reproduction script</summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "pg"
end

require "active_record"
require "active_support/log_subscriber/test_helper"

require "minitest/autorun"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "rails_bug04")
ActiveRecord::Base.logger = nil

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.string :email
  end

  create_table :messages, force: true do |t|
    t.bigint :user_id
    t.string :content
  end
end

class User < ActiveRecord::Base
  has_many :messages
end

class Message < ActiveRecord::Base
  belongs_to :user
end

class BugTest < Minitest::Test
  def setup
    10.times do |n|
      user = User.create(email: "user#{n + 1}@example.com")
      Message.create(user: user, content: "Message #{n + 1}")
    end

    ActiveSupport::LogSubscriber.colorize_logging = false
  end

  def test_where_in_logging_bug
    old_logger = ActiveRecord::Base.logger
    logger = ActiveSupport::LogSubscriber::TestHelper::MockLogger.new
    ActiveRecord::Base.logger = logger

    Message.includes(:user).load

    assert_match(%{["id", 1], ["id", 2], ["id", 3], ["id", 4], ["id", 5], ["id", 6], ["id", 7], ["id", 8], ["id", 9], ["id", 10]}, logger.logged(:debug).last)
  ensure
    ActiveRecord::Base.logger = old_logger
  end
end
```
</details>

eileencodes reviewed (changes requested):
It feels off to me that we're building the bind values where we create the HomogeneousIn object. I'm not sure the correct fix is without spending more time on this. I likely won't get around to digging in more this week. 

Unfortunately this also undoes the performance gain of `HomogeneousIn`.

Before this change:

```
Warming up --------------------------------------
      where with ids    14.000  i/100ms
Calculating -------------------------------------
      where with ids    146.895  (± 3.4%) i/s -    742.000  in   5.057694s
```

After:

```
Warming up --------------------------------------
      where with ids    10.000  i/100ms
Calculating -------------------------------------
      where with ids    117.217  (± 3.4%) i/s -    590.000  in   5.040589s
``` 

Here's the benchmark code:

```ruby
require "active_record"
require "minitest/autorun"
require "stackprof"
require "benchmark/ips"

ActiveRecord::Base.establish_connection(adapter: "mysql2", username: "rails", database: "activerecord_unittest")

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_association_stuff
    ids = (1..1000).each.map do |n|
      Post.create!.id
    end

    Benchmark.ips do |x|
      x.report("where with ids") do
        Post.where(id: ids).to_a
      end
    end
  end
end
```

ricardotk002 commented:
Fair point. I'm also unsure if it's the right solution.

I checked #41136 and it seems like a different approach, although with some corner cases:

```ruby
Developer.where(id: [1, 2, 3, 4, 5], created_at: Time.now).load

# New format
# => SELECT "developers".* FROM "developers" WHERE "developers"."id" IN ($1, $2, $3, $4, $5) AND "developers"."created_at" = $6  [1, 2, 3, 4, 5, ["created_at", "2021-01-22 01:49:07.085656"]]

# Previous format
# => SELECT "developers".* FROM "developers" WHERE "developers"."id" IN ($1, $2, $3, $4, $5) AND "developers"."created_at" = $6  [["id", 1], ["id", 2], ["id", 3], ["id", 4], ["id", 5], ["created_at", "2021-01-22 01:50:47.720448"]]
```

On a side note, I ran my initial benchmarks against the baseline you used (`#where` with sanitized SQL). I tried yours and still the difference seems reasonable. 

Before:

```
Warming up --------------------------------------
      where with ids    10.000  i/100ms
 where with sanitize    17.000  i/100ms
Calculating -------------------------------------
      where with ids    115.849  (± 1.7%) i/s -    580.000  in   5.008217s
 where with sanitize    173.993  (± 3.4%) i/s -    884.000  in   5.088834s

Comparison:
 where with sanitize:      174.0 i/s
      where with ids:      115.8 i/s - 1.50x  (± 0.00) slower
```

After:

```
Warming up --------------------------------------
      where with ids    11.000  i/100ms
 where with sanitize    16.000  i/100ms
Calculating -------------------------------------
      where with ids    116.501  (± 1.7%) i/s -    583.000  in   5.006341s
 where with sanitize    176.277  (± 2.3%) i/s -    896.000  in   5.085439s

Comparison:
 where with sanitize:      176.3 i/s
      where with ids:      116.5 i/s - 1.51x  (± 0.00) slower
```

My guess is that the performance gain comes from not having to visit an `In` node and its left/right leaves. The binds are still visited with `#add_binds`, which works similarly to `#inject_join`.

<details>
  <summary>Benchmark script</summary>

```ruby
require "active_record"
require "minitest/autorun"
require "stackprof"
require "benchmark/ips"

ActiveRecord::Base.establish_connection(adapter: "mysql2", username: "rails", database: "activerecord_unittest")

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_association_stuff
    ids = (1..1000).each.map do |n|
      Post.create!.id
    end

    Benchmark.ips do |x|
      x.report("where with ids") do
        Post.where(id: ids).to_a
      end

      x.report("where with sanitize") do
        Post.where(ActiveRecord::Base.sanitize_sql(["id IN (?)", ids])).to_a
      end

      x.compare!
    end
  end
end
```

</details>

Feel free to close this one if the solution doesn't apply, I understand it's up to you to decide which option is better long term.

ricardotk002 commented:
@eileencodes @kamipo I tried another approach, which is similar (slightly lighter) to the implementation pre-refactor. Performance seems to stay similar using the benchmarks provided above. I also borrowed @kamipo's tests to verify that everything works fine.

The thing you need to determine is if you still want to keep the old logging format (there are some issues that reference this PR):

```
#  SELECT * FROM events WHERE id IN (?, ?)  [["id", 1], ["id", 2]]
```

Or keep the new concise version https://github.com/rails/rails/pull/41136 (related https://github.com/rails/rails/pull/29944)

```
#  SELECT * FROM events WHERE id IN (?, ?)  [1, 2]
```

Which has some curious cases as I mentioned here https://github.com/rails/rails/pull/41068#issuecomment-765066261.

I'm more than happy to try another approach, if you give me some direction. 

Otherwise, I think you should move ahead with the new concise format, I don't want this PR to keep the issue unsolved.

kamipo committed:
Fix binds logging for `HomogeneousIn`

Make bind attribute if the bind collector is used.

Before:

```
User Load (0.5ms)  SELECT "users".* FROM "users" WHERE "users"."id" IN (?, ?, ?, ?, ?)  [[nil, 1], [nil, 2], [nil, 3], [nil, 4], [nil, 5]]
```

After:

```
User Load (0.5ms)  SELECT "users".* FROM "users" WHERE "users"."id" IN (?, ?, ?, ?, ?)  [["id", 1], ["id", 2], ["id", 3], ["id", 4], ["id", 5]]
```

(cherry picked from commit cd1a3705c5f18e431e5311c5dc123ed765752ecf)

kamipo commented:
I'd prefer to make bind attribute only if the bind collector is used, it doesn't have performance penalty if `prepared_statements: false` which is by default for mysql2 adapter.

https://github.com/rails/rails/compare/fix_binds_logging

ricardotk002 commented:
@kamipo Your solution works wonderfully. Do you want me to cherry-pick your commit into this PR? Or do you prefer to open a new one?

ricardotk002 commented:
@eileencodes @kamipo I just went ahead and cherry-picked @kamipo's commit to this branch. Please let me know if there's anything else I can do to consider the issue fixed. Thank you both for your help!

kamipo commented:
I'm merging this since in general changes for the performance should not break any public behavior.

We haven't tested the format of the binds part, but I think losing the attr name from the log for the performance is over optimizing.

Now, changes in this PR makes bind attribute only if the bind collector is used, that is to say it doesn't have performance penalty if `prepared_statements: false` which is by default for mysql2 adapter.
